### ISSUE TITLE
Field selector for finding unscheduled pods: "DesiredState.Host=" or "CurrentState.Host="?

### ISSUE THREAD

**Issue by nikhiljindal** [GoogleCloudPlatform/kubernetes#5207]
Whats the correct field selector for finding unscheduled pods?
We support "DesiredState.Host=" currently, but if I understand correctly, it should be "CurrentState.Host=".

@lavalamp ?

**Comment by lavalamp** [comment]
@bgrant0607 has a strong opinion, IIRC, the scheduler should look at & set only desired state. probably not worth changing this for v1beta1/2 apis.

**Comment by bgrant0607** [comment]
It should be PodSpec in v1beta3, but we need to actually set that field when we bind the pod. Fixing it for v1beta3 should fix it for v1beta1/2.

**Comment by lavalamp** [comment]
(Yes--my "set" referred to writing a binding, of course.)

**Comment by bgrant0607** [comment]
More complete list of steps:
* Write PodSpec.Host when binding
* Run fit predicates wherever PodSpec.Host can be set: create, update, binding
* Kubelet watches PodSpec.Host (rather than boundPods)
* Scheduler watches PodSpec.Host

**Comment by nikhiljindal** [comment]
There are 2 things:
The field that we check internally and the field selector specified by the user.
Right now, the field that we check internally is pod.Status.Host.
The field selector specified by the user is "DesiredState.Host" in v1beta1,2 and Status.Host in v1beta3.

Based on the replies here, it seems that we want the scheduler to look at the spec and not the status.
Hence, the scheduler should query for:
`/v1beta1/pods?fields=DesiredState.Host!=` and
`/v1beta3/pods?fields=Spec.Host!=`
and we should be looking at the Spec.Host field internally
Is that correct?

**Comment by bgrant0607** [comment]
Yes, but the other steps I mentioned are prerequisites for this change.

**Comment by nikhiljindal** [comment]
ok. Thanks for the explanation.
@bgrant0607 : Can you assign this an appropriate priority?

**Comment by bgrant0607** [comment]
This is confusing enough, an important enough API (e.g., for integration with Mesos and YARN), and closely enough related to other goals (v1beta3, elimination of bound pods, elimination of pod cache) that I think it's P1.

**Comment by jszczepkowski** [comment]
I'll work on it.

**Comment by jdef** [comment]
xref https://github.com/mesosphere/kubernetes-mesos/issues/168

**Comment by jszczepkowski** [comment]
@bgrant0607 Can you tell us more about implementation step:
"Run fit predicates wherever PodSpec.Host can be set: create, update, binding"
The other steps are already addressed by our PRs, but we are not sure what this one means. It is some kind of feasibility for binding? If yes, where it is/should be checked?

**Comment by bgrant0607** [comment]
We removed hostPort conflict checking in #4619. The problem was that it was applied when updating BoundPods, which we were trying to remove.

We do plan to replicate apiserver and support multiple, workload- and/or user-specific schedulers. However, unlike [Omega](http://eurosys2013.tudos.org/wp-content/uploads/2013/paper/Schwarzkopf.pdf), I don't think we should enforce invariants at the storage layer. 

We added conflict checking to Kubelet in #5019. In general, the intent is that Kubelet should do admission control and be the final arbiter to ensure its own invariants and to ensure QoS of the running pods under dynamically varying load. 

Other than host ports, I don't know that any of the other fit predicates are urgently needed. We should look at the scheduler's fit predicates, though, and decide which ones, if any, should be enforced by Kubelet:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L35
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/scheduler/predicates.go

The resource and node-selector predicates are obvious candidates.

Ideally, though, pods that failed the fit predicates wouldn't reach Kubelet often, so we should do "best effort" checking earlier. Where exactly is TBD. It can't go into validation, for reasons discussed in #2457.

**Comment by davidopp** [comment]
@davidopp

**Comment by davidopp** [comment]
@jszczepkowski or @nikhiljindal can you summarize what remains to be done for this issue?

**Comment by bgrant0607** [comment]
@davidopp Take a look at #5501 if you like.

**Comment by davidopp** [comment]
Can you explain how #5501 is related to this issue? Brian listed four things here
https://github.com/GoogleCloudPlatform/kubernetes/issues/5207#issuecomment-77969711
but that doesn't seem to correspond to any of them (I may be missing something though).

**Comment by bgrant0607** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/issues/5207#issuecomment-78613182

**Comment by jszczepkowski** [comment]
@davidopp Out of four steps listed, three are already done. The only step missing is running fit predicates on kubelet. There is an open PR for it #5501, on which I'm working on: resource predicates is already implemented and I'm now working on node selector.

**Comment by davidopp** [comment]
OK, thanks, please make sure to close this issue when you finish #5501.
